### PR TITLE
allow quantum tanks to lock their fluids

### PR DESCRIPTION
**Pull request by TechLord22**

### CODE REVIEW

**Comment by DStrand1** [comment]
Stealing this PR to improve it a bit and add more features :trollface:

**Comment by DStrand1** [comment]
Changes since my commits:
- Quantum Tank now works as a FluidHandlerItemStack, meaning it can be used to empty into inventories (like other Quantum Tanks, Fluid Hatches, etc), and can be used to click into GT machine fluid slots
- Locking now forces the Fluid to be rendered in the Fluid Slot, even when empty (additionally keeps the name/amount in the UI)
- You can now use JEI ghost ingredients to set the locked fluid
- You can also use other FluidHandlerItemStacks to set the locked fluid
- There are two new buttons related to voiding:
    - Void (All): Void all fluids of the tank
    - Void (Partial): Void fluids in the tank only when the tank is 95% full

**Comment by DStrand1** [comment]
Changing this to `Release Priority` as this PR now includes several breaking changes to the Tank's NBT

File: src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+                    phantomTank.setFluid(stack);
+                }
+            }
+        });
+    }
+
+    @Override
+    public void handleClientAction(int id, PacketBuffer buf) {
+        super.handleClientAction(id, buf);
+        if (id == 12) {
```
Comment (TechLord22) [inline src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java:89]: Should put this value in GregtechDataCodes.

File: src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+                stack.setCount(1);
+                IFluidHandlerItem fluidHandler = stack.getCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY, null);
+                if (fluidHandler != null) {
+                    FluidStack resultStack = fluidHandler.drain(Integer.MAX_VALUE, false);
+                    phantomTank.setFluid(resultStack);
+                }
+            } else {
+                phantomTank.setFluid(null);
+            }
+        } else if (id == 13) {
```
Comment (TechLord22) [inline src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java:101]: Should put this value in GregtechDataCodes.

File: src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        if (isMouseOverElement(mouseX, mouseY)) {
+            return lastFluidInTank == null ? lastPhantomStack : lastFluidInTank;
+        }
+        return null;
+    }
+
+    @Override
+    public boolean mouseClicked(int mouseX, int mouseY, int button) {
+        if (isMouseOverElement(mouseX, mouseY)) {
+            writeClientAction(12, buf -> {});
```
Comment (TechLord22) [inline src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java:123]: Should put this value in GregtechDataCodes.

File: src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        }
+    }
+
+    @Override
+    public void detectAndSendChanges() {
+        super.detectAndSendChanges();
+        FluidStack stack = phantomTank.getFluid();
+        if (stack == null && lastPhantomStack != null) {
+            lastPhantomStack = null;
+            writeUpdateInfo(10, buf -> {});
```
Comment (TechLord22) [inline src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java:193]: Should put this value in GregtechDataCodes.

File: src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        } else if (stack != null) {
+            if (!stack.isFluidEqual(lastPhantomStack)) {
+                lastPhantomStack = stack.copy();
+                NBTTagCompound stackTag = stack.writeToNBT(new NBTTagCompound());
+                writeUpdateInfo(11, buf -> buf.writeCompoundTag(stackTag));
```
Comment (TechLord22) [inline src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java:198]: Should put this value in GregtechDataCodes.

File: src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+            } else if (stack.amount != 0) {
+                lastPhantomStack.amount = 0;
+                writeUpdateInfo(12, buf -> {});
```
Comment (TechLord22) [inline src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java:201]: Should put this value in GregtechDataCodes.

File: src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+            }
+        }
+    }
+
+    @Override
+    public void readUpdateInfo(int id, PacketBuffer buf) {
+        super.readUpdateInfo(id, buf);
+        if (id == 10) {
```
Comment (TechLord22) [inline src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java:209]: Should put this value in GregtechDataCodes.

File: src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+            lastPhantomStack = null;
+        } else if (id == 11) {
```
Comment (TechLord22) [inline src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java:211]: Should put this value in GregtechDataCodes.

File: src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+            NBTTagCompound stackTag;
+            try {
+                stackTag = buf.readCompoundTag();
+            } catch (IOException ignored) {
+                return;
+            }
+            lastPhantomStack = FluidStack.loadFluidStackFromNBT(stackTag);
+        } else if (id == 12) {
```
Comment (TechLord22) [inline src/main/java/gregtech/api/gui/widgets/PhantomTankWidget.java:219]: Should put this value in GregtechDataCodes.

**Review by bruberu — APPROVED** [review]
This also works properly on a server.

**Pull request merged**